login/login.c

```diff
@@ -243,33 +243,45 @@ void login_syslog(glome_login_config_t* config, pam_handle_t* pamh,
   }
 }
 
-// read_stdin reads printable characters from stdin into buf. It returns:
+// read_stdin reads characters from stdin into buf. It returns:
 // -1, if it encounters an error while reading
-// -2, if it encounters EOF
 // (buflen-1) if it read buflen-1 characters
 // <(buflen-1), if a newline was read before the buffer was full
 // If the return value is >=0, the buf is NULL-terminated.
 static int read_stdin(char* buf, int buflen) {
-  int i = 0;
+  // Return error if we got no characters.
+  if (fgets(buf, buflen, stdin) == NULL) {
+    perror("ERROR when reading from stdin");
+    return -1;
+  }
+
+  bool newline = false;
+  int len = strlen(buf);
+  if (buf[len - 1] == '\n') {
+    newline = true;
+    buf[len - 1] = '\0';
+    len--;
+  }
 
-  while (i < buflen - 1) {
-    int n = read(STDIN_FILENO, buf + i, 1);
-    if (n < 0) {  // error while reading from stdin
-      perror("ERROR when reading from stdin");
+  // Return error if we got a non-printable character.
+  for (int i = 0; i < len; i++) {
+    if (buf[i] < 0x20 || buf[i] > 0x7e) {
+      perror("ERROR invalid characters read from stdin");
```

Comment:
This should not use `perror`, but `errorf`.

Reply:
Done.